### PR TITLE
Generate block flips

### PR DESCRIPTION
**Pull request by jcolemang**
#2 
https://github.com/orgs/RHIT-XPrize/projects/1#card-6573891

### CODE REVIEW

File: generate_instructions_test.py [hunk 1]
```diff
@@ -258,3 +252,16 @@ def test_solve_board_final_action_yields_goal(self):
...
+        letters = string.ascii_uppercase
+        colors = ['Red', 'Green', 'Blue', 'Gold', 'Brown', 'Purple', 'Yellow', 'Pink']
+
+        start_config = gi.random_configuration(num_blocks, letters, colors)
+        final_config = start_config.scatter().mark_complete()
+
+        actions = gi.solve_board(start_config, final_config)
+        solved_config = actions[len(actions) - 1].end_conf
+
+        assert final_config == solved_config
```
Comment (jcolemang) [inline generate_instructions_test.py]: I'm second guessing this a little considering how long CI is suddenly taking. On the other hand, I don't think we need super low latency for the PR test turnarounds.

File: generate_instructions.py [hunk 2]
```diff
@@ -32,10 +32,13 @@ def __init__(self, side1_letter,
         self.block_id = block_id if block_id > 0 else Block.get_next_block_id()
 
     def __eq__(self, other):
-        return self.block_id == other.block_id
+        return type(other) == type(self) and self.block_id == other.block_id
 
-    def __ne__(self, other):
-        return not self == other
```
Comment (Lewis-Kelley) [inline generate_instructions.py]: Why delete `__ne__`? I can only see that becoming a subtle problem in the future.

File: generate_instructions.py [hunk 3]
```diff
@@ -81,28 +94,37 @@ def __init__(self, blocks, final_blocks=[]):
...
 
-    @staticmethod
-    def get_instruction(moved_block):
-        point = moved_block.position
-        phrase = 'move the %s %s block here' \
-                 % (moved_block.side1_color,
-                    moved_block.side1_letter)
-        return Instruction(phrase, point)
-
     def generate_action(self, goal_config):
```
Comment (Lewis-Kelley) [inline generate_instructions.py]: This method is getting a little large. Consider splitting some functionality out, such as creating the next configuration.

File: generate_instructions.py [hunk 4]
```diff
@@ -116,6 +138,26 @@ def __str__(self):
     def mark_complete(self):
         return Configuration([], self.current_blocks + self.final_blocks)
 
+    def __eq__(self, other):
```
Comment (Lewis-Kelley) [inline generate_instructions.py]: Doesn't Python have a concept of a `set`? Could we just use that to hold the blocks and use the existing `Block.__eq__` method to assert the configurations are equal? If so, we could remove most of the body of this method.

File: generate_instructions_test.py (hunk 1 above)
Comment (Lewis-Kelley) [reply generate_instructions_test.py]: I'm not sure if this is the problem... The builds seem to have just not started. Maybe their servers are down or something?

File: generate_instructions.py (hunk 2 above)
Comment (kgroble) [reply generate_instructions.py]: I believe in the absence of `__ne__`, inequality checks become `not __eq__` - why risk `__eq__` and `__ne__` getting out of sync?

Comment (Lewis-Kelley) [reply generate_instructions.py]: https://stackoverflow.com/questions/4352244/python-should-i-implement-ne-operator-based-on-eq

Comment (jcolemang) [reply generate_instructions.py]: Yeah he pointed that out to me last week or something.

File: generate_instructions.py (hunk 4 above)
Comment (jcolemang) [reply generate_instructions.py]: It does, you just write something like `set([1, 2, 3])`. However, checking if two sets are equal would check that both sets have `__eq__` elements. I don't think that two blocks should require that they appear the same to be `__eq__`, so I can't do `set(a) == set(b)` unless I swapped out the `__eq__` method at runtime which seems like a bad idea. Sets would still be reasonable, but I can't find a way to get a specific element out of the set to check that they look the same given that the are `__eq__`.

All that being said, looking at it now I think I could use a dictionary to make it faster and easier to read.

File: generate_instructions_test.py [hunk 5]
```diff
@@ -12,6 +13,12 @@ def test_block_equality_with_given_ids():
     assert gi.Block('A', 'BLUE', 'B', 'GREEN', (0, 0), 30) \
         == gi.Block('D', 'YELLOW', 'C', 'ORANGE', (-1, 2), 30)
 
+def test_block_looks_the_same():
```
[the remark this answers is not in the document]
Comment (jcolemang) [reply generate_instructions_test.py]: Good call yeah I'll do that

File: generate_instructions.py (hunk 4 above)
Comment (jcolemang) [reply generate_instructions.py]: It looks much nicer now

File: generate_instructions.py [hunk 6]
```diff
@@ -139,24 +149,11 @@ def mark_complete(self):
...
-                if not b.looks_the_same(other):
-                    return False
-            return True
-
-        # a is a subset of b and b is a subset of a
-        return block_list_subset(self.current_blocks, other.current_blocks) and \
-            block_list_subset(other.current_blocks, self.current_blocks) and \
-            block_list_subset(self.final_blocks, other.final_blocks) and \
-            block_list_subset(other.final_blocks, self.final_blocks)
+        eq_helper = lambda ls1, ls2: all([x == y and x.looks_the_same(y) \
```
Comment (Lewis-Kelley) [inline generate_instructions.py]: I like it!

File: generate_instructions_test.py (hunk 1 above)
Comment (jcolemang) [reply generate_instructions_test.py]: Yeah I don't think this is actually a problem

**Pull request merged**